### DISCUSSION TITLE
Zitadel setup guide

### DISCUSSION THREAD

**Discussion by aldervall** [9p4/jellyfin-plugin-sso#205] [General]
# SSO Login Guide for Jellyfin with ZITADEL

## 1. Create a New Project in ZITADEL

1. **Login to your ZITADEL instance.**
2. **Create a new project** and name it whatever you like.

## 2. Configure the Project for Code Flow

 **Choose `CODE`** as the flow type.

![image](https://github.com/user-attachments/assets/c1bb9b0e-d719-4435-b89f-f6aa48504369)

### Redirect URIs
Add the following Redirect URIs:

- `https://jellyfin.YOURSERVER.COM/sso/OID/r/zitadel`
- `https://jellyfin.YOURSERVER.COM/sso/OID/redirect/zitadel`

### Post Logout URIs
Add the following Post Logout URI:

- `https://jellyfin.YOURSERVER.COM`

 **Press `Continue` and `Create`.**

## 3. Copy Credentials

- **Copy your `Client Secret` and `Client ID` for later use.**

## 4. Configure Roles in ZITADEL

1. **Mark `Assert Roles on Authentication`.**
2. **Mark `Check authorization on Authentication` and press `Save`.**
3. **Go to the Roles section.**

![image](https://github.com/user-attachments/assets/7c11750f-0892-4244-a7f9-343292ac5576)

### Create Groups
Create the following three groups:

- `jellyfin_user`
- `jellyfin_tv`
- `jellyfin_admin`

![image](https://github.com/user-attachments/assets/3551a93b-58fa-492e-9cad-6fdaeb7f602d)

. **Go to your Organization and Copy your Organization's Resource ID** from ZITADEL.

![image](https://github.com/user-attachments/assets/f1ae6584-6307-405e-ad27-d137ad20586f)

## 5. Set Up SSO in Jellyfin

1. **Go to Jellyfin** and download/setup the SSO plugin.
2. **Add the following configuration**:

OID Endpoint:
- `https://zitadel.YOURSERVER.COM/.well-known/openid-configuration`

![image](https://github.com/user-attachments/assets/d18ae6d3-ff9c-42de-b2eb-25d14f225b0c)`

OpenID Client ID:
- Add your `Client ID` from Zitadel

OID Secret: 
- Add your `Client Secret` from Zitadel

### Example Role Mappings
Here are example role mappings to use in your configuration:
Change it to your organization: `{"jellyfin_tv":{"{ORGANIZATIONID":"{PRIMARYDOMAIN.COM}"}}`

```json
{"jellyfin_user":{"265153045849972739":"{demo-vendor.com}"}}
{"jellyfin_tv":{"265153045849972739":"{demo-vendor.com}"}}
{"jellyfin_admin":{"265153045849972739":"{demo-vendor.com}"}}
   ```
   
### Admin Roles:
Change it to your organization: `{"jellyfin_admin":{"{ORGANIZATIONID":"{PRIMARYDOMAIN.COM}"}}`

```json
{"jellyfin_admin":{"265153045849972739":"{demo-vendor.com}"}}
   ```
![image](https://github.com/user-attachments/assets/bf0d51f8-1a53-4f60-9e6a-e6a336808c64)

### Live TV Roles:

```json
{"jellyfin_tv":{"265153045849972739":"{demo-vendor.com}"}}
{"jellyfin_admin":{"265153045849972739":"{demo-vendor.com}"}}
   ```
### Live TV Management Roles:
Change it to your organization: `{"jellyfin_tv":{"{ORGANIZATIONID":"{PRIMARYDOMAIN.COM}"}}`
```json
{"jellyfin_tv":{"265153045849972739":"{demo-vendor.com}"}}
{"jellyfin_admin":{"265153045849972739":"{demo-vendor.com}"}}
   ```
### Scopes and Claims
Add the following scopes and claims:

![image](https://github.com/user-attachments/assets/7a677b29-9f8f-44fb-aeb8-65ff1f0341d3)

- **Role Claims**:
 ``` 
 urn:zitadel:iam:org:project:{projectResourceId}:roles
 ```
- **Scopes**:
 ``` 
openid
email
profile
urn:zitadel:iam:org:project:id:zitadel:aud
urn:zitadel:iam:org:project:{projectResourceId}:roles
 ```

- **Set default Provider**:

 ``` 
Jellyfin.Server.Implementations.Users.DefaultAuthenticationProvider
 ```
- **Set default username claim**: 

```
preferred_username
 ```
 
- **Scheme Override**:
I'm configuring mine to use HTTPS, but your setup might be different.
``` 
 https
```

 **Save** the configuration**:

## 6. Add Users in ZITADEL

- **Add users** to the project to authenticate and ad them to a role.
### I've encountered issues when assigning two roles to a single user in Zitadel. Jellyfin will responds with a 'Permission Denied' error.

![image](https://github.com/user-attachments/assets/8b5fa68a-d337-4c15-a55b-0bfce881d2ee)

## 7. Test the SSO Setup

1. **Go to**:
```
https://jellyfin.YOURSERVER.com/sso/OID/start/zitadel
```
2. **Try to login** with SSO to verify the setup.

**Comment by 9p4** [comment]
If you're willing, would you like to create a PR to add this to the official docs? https://github.com/9p4/jellyfin-plugin-sso/blob/main/providers.md

**Comment by ilbarone87** [comment]
@aldervall thank you very much for this guide! I've tried to set it up too with zitadel but I seem to be stuck on error "Error. Check permissions." when trying to access `https://jellyfin.mydomain.net/sso/OID/start/zitadel`

This is my xml:

```
<?xml version="1.0" encoding="utf-8"?>
<PluginConfiguration xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <SamlConfigs />
  <OidConfigs>
    <item>
      <key>
        <string>zitadel</string>
      </key>
      <value>
        <PluginConfiguration>
          <OidEndpoint>https://zitadel.mydomain.net/.well-known/openid-configuration</OidEndpoint>
          <OidClientId>291603344764502395</OidClientId>
          <OidSecret>SUPERSECRETDOMAIN</OidSecret>
          <Enabled>true</Enabled>
          <EnableAuthorization>true</EnableAuthorization>
          <EnableAllFolders>true</EnableAllFolders>
          <EnabledFolders />
          <AdminRoles>
            <string>{"jellyfin_admin":{"259430593291419955":"{myorg.zitadel.mydomain.net}"}}</string>
          </AdminRoles>
          <Roles>
            <string>{"jellyfin_admin":{"259430593291419955":"{myorg.zitadel.mydomain.net}"}}</string>
          </Roles>
          <EnableFolderRoles>false</EnableFolderRoles>
          <EnableLiveTvRoles>false</EnableLiveTvRoles>
          <EnableLiveTv>false</EnableLiveTv>
          <EnableLiveTvManagement>false</EnableLiveTvManagement>
          <LiveTvRoles>
            <string>{"jellyfin_admin":{"259430593291419955":"{myorg.zitadel.mydomain.net}"}}</string>
          </LiveTvRoles>
          <LiveTvManagementRoles>
            <string>{"jellyfin_admin":{"259430593291419955":"{myorg.zitadel.mydomain.net}"}}</string>
          </LiveTvManagementRoles>
          <FolderRoleMappings />
          <RoleClaim>urn:zitadel:iam:org:project:{291603061934129531}:roles</RoleClaim>
          <OidScopes>
            <string>openid</string>
            <string>email</string>
            <string>profile</string>
            <string>urn:zitadel:iam:org:project:id:zitadel:aud</string>
            <string>urn:zitadel:iam:org:project:{291603061934129531}:roles</string>
          </OidScopes>
          <DefaultProvider>Jellyfin.Server.Implementations.Users.DefaultAuthenticationProvider</DefaultProvider>
          <SchemeOverride>https</SchemeOverride>
          <NewPath>true</NewPath>
          <CanonicalLinks />
          <DefaultUsernameClaim>preferred_username</DefaultUsernameClaim>
          <DisableHttps>false</DisableHttps>
          <DoNotValidateEndpoints>false</DoNotValidateEndpoints>
          <DoNotValidateIssuerName>false</DoNotValidateIssuerName>
        </PluginConfiguration>
      </value>
    </item>
  </OidConfigs>
</PluginConfiguration>
```

![image](https://github.com/user-attachments/assets/445f5db1-8b53-4bb0-a3d1-231ef3f39554)

![image](https://github.com/user-attachments/assets/eb820565-2c3e-4e52-89e6-17989ae228eb)

Log from Jellyfin
`[2024-10-30 17:26:07.289 +00:00] [WRN] OpenID user "259430593291485491" has one or more incorrect role claims: [{ Type: "amr", Value: "user" }, { Type: "amr", Value: "mfa" }, { Type: "azp", Value: "291603344764502395" }, { Type: "client_id", Value: "291603344764502395" }, { Type: "email", Value: "zitadel-admin-sa@zitadel.mydomain.net" }, { Type: "email_verified", Value: "True" }, { Type: "family_name", Value: "admin" }, { Type: "gender", Value: "male" }, { Type: "given_name", Value: "admin" }, { Type: "groups", Value: "jellyfin_admin" }, { Type: "locale", Value: "en" }, { Type: "name", Value: "admin admin" }, { Type: "nickname", Value: "nickname" }, { Type: "preferred_username", Value: "zitadel-admin-sa@zitadel.mydomain.net" }, { Type: "sid", Value: "V1_291600807009190267" }, { Type: "sub", Value: "259430593291485491" }, { Type: "updated_at", Value: "1711129531" }, { Type: "urn:zitadel:iam:org:project:291603061934129531:roles", Value: "{\"jellyfin_admin\":{\"259430593291419955\":\"myorg.zitadel.mydomain.net\"}}" }, { Type: "urn:zitadel:iam:org:project:roles", Value: "{\"jellyfin_admin\":{\"259430593291419955\":\"myorg.zitadel.mydomain.net\"}}" }, { Type: "email_verified", Value: "true" }]. Expected any one of: ["{\"jellyfin_admin\":{\"259430593291419955\":\"{myorg.zitadel.mydomain.net}\"}}"]`

Something to notice: if I remove the Roles mapping I can login without issue (user doesn't have any permission and that's expected) but still is a problem since manually given permissions just get overwritten every log in.

**Reply by aldervall** [reply]
Hm, 
Looks like it does not like the formatting of the users?
I had some issues with this. 

Have you only added your domain in the roles? and not "myorg.zitadel.domain.com"?
Like this
{"jellyfin_admin":{"264738154777477122":"domain.com"}}

**Reply by ilbarone87** [reply]
I've used the primary domain as it is in Organization tab in Zitadel mgmt console "myorg.zitadel.domain.com". 
I've also added a secondary domain to my Organization, that consists in "domain.com" and used this one in the role mapping, but still same error.

**Reply by aldervall** [reply]
Under 
Role Claim and Request Additional Scopes try and remove {} in projectResourceId.
urn:zitadel:iam:org:project:{291603061934129531}:roles

So it looks like this:
`urn:zitadel:iam:org:project:291603061934129531:roles`

**Reply by ilbarone87** [reply]
Still same.
Have also added another domain and  tried {"jellyfin_admin":{"{ORGANIZATIONID":"{PRIMARYDOMAIN.COM}"}} with zitadel.mydomain.net. Still same error

**Reply by ilbarone87** [reply]
Do you mind dropping me your SSO-Auth.xml (cleaned from your personal details) so i can compare it please?

**Reply by aldervall** [reply]
```
<?xml version="1.0" encoding="utf-8"?>
<PluginConfiguration xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <SamlConfigs />
  <OidConfigs>
    <item>
      <key>
        <string>zitadel</string>
      </key>
      <value>
        <PluginConfiguration>
          <OidEndpoint>https://example.com/.well-known/openid-configuration</OidEndpoint>
          <OidClientId>CLIENT_ID_PLACEHOLDER</OidClientId>
          <OidSecret>SECRET_PLACEHOLDER</OidSecret>
          <Enabled>true</Enabled>
          <EnableAuthorization>true</EnableAuthorization>
          <EnableAllFolders>true</EnableAllFolders>
          <EnabledFolders />
          <AdminRoles>
            <string>{"jellyfin_admin":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
          </AdminRoles>
          <Roles>
            <string>{"jellyfin_user":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
            <string>{"jellyfin_tv":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
            <string>{"jellyfin_admin":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
          </Roles>
          <EnableFolderRoles>false</EnableFolderRoles>
          <EnableLiveTvRoles>true</EnableLiveTvRoles>
          <EnableLiveTv>false</EnableLiveTv>
          <EnableLiveTvManagement>false</EnableLiveTvManagement>
          <LiveTvRoles>
            <string>{"jellyfin_tv":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
            <string>{"jellyfin_admin":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
          </LiveTvRoles>
          <LiveTvManagementRoles>
            <string>{"jellyfin_admin":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
            <string>{"jellyfin_tv":{"ROLE_ID_PLACEHOLDER":"example.com"}}</string>
          </LiveTvManagementRoles>
          <FolderRoleMappings />
          <RoleClaim>urn:zitadel:roles</RoleClaim>
          <OidScopes>
            <string>openid</string>
            <string>email</string>
            <string>profile</string>
            <string>urn:zitadel:aud</string>
            <string>urn:zitadel:roles</string>
          </OidScopes>
          <DefaultProvider>Default.AuthenticationProvider</DefaultProvider>
          <SchemeOverride>https</SchemeOverride>
          <NewPath>true</NewPath>
          <CanonicalLinks>
            <item>
              <key>
                <string>test</string>
              </key>
              <value>
                <guid>GUID_PLACEHOLDER_1</guid>
              </value>
            </item>
            <item>
              <key>
                <string>example@example.com</string>
              </key>
              <value>
                <guid>GUID_PLACEHOLDER_2</guid>
              </value>
            </item>
            <item>
              <key>
                <string>username</string>
              </key>
              <value>
                <guid>GUID_PLACEHOLDER_3</guid>
              </value>
            </item>
            <item>
              <key>
                <string>another@example.com</string>
              </key>
              <value>
                <guid>GUID_PLACEHOLDER_4</guid>
              </value>
            </item>
            <item>
              <key>
                <string>user@example.com</string>
              </key>
              <value>
                <guid>GUID_PLACEHOLDER_5</guid>
              </value>
            </item>
          </CanonicalLinks>
          <DefaultUsernameClaim>preferred_username</DefaultUsernameClaim>
          <DisableHttps>false</DisableHttps>
          <DoNotValidateEndpoints>false</DoNotValidateEndpoints>
          <DoNotValidateIssuerName>false</DoNotValidateIssuerName>
        </PluginConfiguration>
      </value>
    </item>
  </OidConfigs>
</PluginConfiguration>

```

**Reply by ilbarone87** [reply]
minimal differences with mine. Have edited to match yours btw, still not working

**Reply by areitz86** [reply]
just tried the walkthrough. got the same results "Error. Check permissions."

**Reply by Tothal1** [reply]
If anyone still has the Error. Check permissions problem.
I managed to fix it by changing from
```
{"jellyfin_user":{"265153045849972739":"{demo-vendor.com}"}}
{"jellyfin_tv":{"265153045849972739":"{demo-vendor.com}"}}
{"jellyfin_admin":{"265153045849972739":"{demo-vendor.com}"}}
```
to
```
{"jellyfin_user":{"265153045849972739":"demo-vendor.com"}}
{"jellyfin_tv":{"265153045849972739":"demo-vendor.com"}}
{"jellyfin_admin":{"265153045849972739":"demo-vendor.com"}}
```
removing the brackets around the domain made it work

**Reply by bannert1337** [reply]
> Under Role Claim and Request Additional Scopes try and remove {} in projectResourceId. urn:zitadel:iam:org:project:{291603061934129531}:roles
> 
> So it looks like this: `urn:zitadel:iam:org:project:291603061934129531:roles`

For anyone coming across this:
Even after applying the change suggested by @Tothal1 in https://github.com/9p4/jellyfin-plugin-sso/discussions/205#discussioncomment-12531143 the same error appeared.

The comment of @aldervall was hidden, but contained the right answer to fix the problem. Now it works!

**Comment by michaelkuty** [comment]
## Solution for Zitadel Cross-Organization Role Claims

For anyone struggling with Zitadel multi-org setups where the `groups` claim approach doesn't work for cross-organization users (due to [zitadel/zitadel#9141](https://github.com/zitadel/zitadel/issues/9141) — Actions don't execute for cross-org users), I've created a fork with a small patch that adds support for JSON object keys as role names.

### The Problem

Zitadel encodes roles as JSON object keys:
```json
"urn:zitadel:iam:org:project:{projectId}:roles": {
  "jellyfin_admin": {"org_id": "org_domain"},
  "jellyfin_user": {"org_id": "org_domain"}
}
```

The plugin only supports raw strings or JSON arrays — not object keys. The workaround using Zitadel Actions to flatten roles into a `groups` claim fails for users from other organizations.

### The Fix

The patched plugin parses JSON object keys as role names when the claim value is a JSON object. This means you can configure:

- **Role Claim:** `urn:zitadel:iam:org:project:{projectId}:roles`
- **Roles:** `jellyfin_user` / `jellyfin_admin`
- **Admin Roles:** `jellyfin_admin`

No Zitadel Actions needed. Works for all organizations.

### Install

Add this plugin repository in Jellyfin (Admin → Plugins → Repositories):
```
https://raw.githubusercontent.com/michaelkuty/jellyfin-plugin-sso/manifest-release/manifest.json
```

Source: https://github.com/michaelkuty/jellyfin-plugin-sso/tree/feat/support-json-object-role-claims

Backward compatible — existing setups (Authelia `groups`, Keycloak `realm_access.roles`) continue to work unchanged.
